File: src/solo.rs
use crate::{block::Block, tx::Tx};
use colored::*;
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Serialize, Deserialize)]
struct NodeResp {
    pub block: Block,
    pub miner_reward: u32,
}

async fn get_block(node_uri: &str) -> String {
    reqwest::get(format!("{}/mine", node_uri))
        .await
        .unwrap()
        .text()
        .await
        .unwrap()
}

async fn send_block(block: &Block, node_uri: &str) -> Result<reqwest::Response, reqwest::Error> {
    reqwest::Client::new()
        .post(format!("{}/mine", node_uri))
        .header("Content-Type", "application/json")
        .body(serde_json::to_string(&block).unwrap())
        .send()
        .await
}

pub async fn solo_mine(node_uri: &str, miner_address: &str) {
    let mut show_no_tx_to_mine = true;

    loop {
        let resp = get_block(node_uri).await;

        if resp == "No transactions to mine" {
            if show_no_tx_to_mine {
                println!(
                    "{}",
                    "No transactions to mine, waiting for 10 seconds..."
                        .to_string()
                        .bright_black(),
                );
                show_no_tx_to_mine = false;
            }
            tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
            continue;
        }

        println!("mining...");
        let start_time = SystemTime::now();

        let node_resp: NodeResp = serde_json::from_str(&resp).unwrap();

        let mut block = node_resp.block.clone();
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let coin_base_tx = Tx {
            sender: "".to_string(),
            receiver: miner_address.to_string(),
            amount: node_resp.miner_reward,
            signature: "".to_string(),
            hash: "".to_string(),
            timestamp,
        };
        block.data.push(coin_base_tx);
        block.merkle_root = block.calculate_merkle_root();
        block.mine();

        println!();
        println!("Block mined: {}", block.hash.bright_green());
        println!("Nonce: {}", block.nonce.to_string().bright_red());
        println!(
            "Time taken to mine: {} seconds",
            SystemTime::now()
                .duration_since(start_time)
                .unwrap()
                .as_secs()
                .to_string()
                .bright_yellow()
        );

        show_no_tx_to_mine = true;

        for _ in 0..10 {
            let resp = send_block(&block, node_uri).await;
            if resp.is_ok() {
                break;
            }
            tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
        }

        println!("Block sent to node");
    }
}

File: mine/solo.go
package mine

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/adibfahimi/pixelpay-miner/core"
	pcore "github.com/adibfahimi/pixelpay-node/core"
)

type getBlockResponse struct {
	Message string       `json:"message"`
	Data    getBlockData `json:"data"`
}

type getBlockData struct {
	Block      pcore.Block `json:"block"`
	MineReward uint        `json:"reward"`
}

func getBlock(nodeAddress string) getBlockResponse {
	url := fmt.Sprintf("%s/mine", nodeAddress)
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		log.Fatalf("client: could not create request: %s\n", err)
	}

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Fatalf("client: could not send request: %s\n", err)
	}

	defer res.Body.Close()

	var data getBlockResponse

	if err := json.NewDecoder(res.Body).Decode(&data); err != nil {
		log.Fatalf("client: could not decode response: %s\n", err)
	}

	return data
}

func sendBlock(block pcore.Block, nodeAddress string) error {
	json, err := json.Marshal(block)
	if err != nil {
		log.Println(err)
	}

	url := fmt.Sprintf("%s/mine", nodeAddress)
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(json))
	req.Header.Set("Content-Type", "application/json")
	if err != nil {
		return fmt.Errorf("client: could not create request: %s", err)
	}

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("client: could not send request: %s", err)
	}

	if res.StatusCode != http.StatusOK {
		return fmt.Errorf("client: unexpected status code: %d", res.StatusCode)
	}

	return nil
}

func MineSolo(nodeAddress string) {
	log.Println("Mining block ...")
	w := core.LoadWallet()

	for {
		log.Println("Getting block ...")
		d := getBlock(nodeAddress)
		log.Println("Block received.")

		if d.Message == "no pending tx" {
			log.Println("No pending tx. Waiting for 15 seconds ...")
			time.Sleep(15 * time.Second)
		} else {
			coinBaseTx := pcore.Tx{
				From:      "",
				To:        w.Address,
				Signature: "",
				Hash:      "",
				Amount:    d.Data.MineReward,
				Timestamp: uint(time.Now().Unix()),
			}

			coinBaseTx.Hash = coinBaseTx.CalculateHash()
			d.Data.Block.Txs = append(d.Data.Block.Txs, coinBaseTx)
			d.Data.Block.MerkleRoot = d.Data.Block.CalculateMerkleRoot()

			log.Println("Mining block ...")
			d.Data.Block.MineBlock()
			log.Println("Block mined. With nonce:", d.Data.Block.Nonce)

			if err := sendBlock(d.Data.Block, nodeAddress); err != nil {
				log.Fatalf("client: could not send block: %s\n", err)
			}

			log.Println("Block sent. mining next block ...")
		}
	}
}
